tests/__init__.py

```diff
@@ -1,4 +1,261 @@
+            'AWS_DEFAULT_REGION': 'us-west-2',
+            'AWS_ACCESS_KEY_ID': 'access_key',
+            'AWS_SECRET_ACCESS_KEY': 'secret_key',
+            'AWS_CONFIG_FILE': '',
+            'AWS_SHARED_CREDENTIALS_FILE': '',
+        }
+
+    def _do_run(self, cmdline):
+        driver = create_clidriver()
+        self._session_stubber.register(driver.session)
+        rc = driver.main(cmdline)
+        self._session_stubber.assert_no_remaining_responses()
+        runner_result = CLIRunnerResult(rc)
+        runner_result.aws_requests = copy.copy(
+            self._session_stubber.received_aws_requests
+        )
+        return runner_result
+
+
+class SessionStubber(object):
+    def __init__(self):
+        self.received_aws_requests = []
+        self._responses = collections.deque()
+
+    def register(self, session):
+        events = session.get_component('event_emitter')
+        events.register_first(
+            'before-parameter-build.*.*', self._capture_aws_request,
+        )
+        events.register_last(
+            'request-created', self._capture_http_request
+        )
+        events.register_first(
+            'before-send.*.*', self._return_queued_http_response,
+        )
+
+    def add_response(self, response):
+        self._responses.append(response)
+
+    def assert_no_remaining_responses(self):
+        if len(self._responses) != 0:
+            raise AssertionError(
+                "The following queued responses are remaining: %s" %
+                self._responses
+            )
+
+    def _capture_aws_request(self, params, model, context, **kwargs):
+        aws_request = AWSRequest(
+            service_name=model.service_model.service_name,
+            operation_name=model.name,
+            params=params,
+        )
+        self.received_aws_requests.append(aws_request)
+        context['current_aws_request'] = aws_request
+
+    def _capture_http_request(self, request, **kwargs):
+        request.context['current_aws_request'].http_requests.append(
+            HTTPRequest(
+                method=request.method,
+                url=request.url,
+                headers=request.headers,
+                body=request.body,
+            )
+        )
+
+    def _return_queued_http_response(self, request, **kwargs):
+        response = self._responses.popleft()
+        return response.on_http_request_sent(request)
+
+
+class BaseResponse(object):
+    def on_http_request_sent(self, request):
+        raise NotImplementedError('on_http_request_sent')
+
+
+class AWSResponse(BaseResponse):
+    def __init__(self, service_name, operation_name, parsed_response):
+        self._service_name = service_name
+        self._operation_name = operation_name
+        self._parsed_response = parsed_response
+        self._service_model = self._get_service_model()
+        self._operation_model = self._service_model.operation_model(
+            self._operation_name)
+        self._validate_parsed_response()
```

Comment:
Doing validation logic in `__init__` is a bit weird, I can see the reason, being defensive against test-writers. Let's add a `validate=True` kwarg here so it can be disabled if needed.